polygeist/lib/polygeist/Passes/ConvertPolygeistToLLVM.cpp

```diff
@@ -90,84 +84,131 @@ struct SubIndexOpLowering : public ConvertOpToLLVMPattern<SubIndexOp> {
-      auto gep = rewriter.create<LLVM::GEPOp>(loc, elemPtrTy, prev, idxs);
-      MemRefDescriptor nexRef = createMemRefDescriptor(
-          loc, viewMemRefType, gep, gep, sizes, strides, rewriter);
-
-      rewriter.replaceOp(subViewOp, {nexRef});
+    Type sourceElemType = sourceMemRefType.getElementType();
+    Type convSourceElemType = getTypeConverter()->convertType(sourceElemType);
+    Type viewElemType = viewMemRefType.getElementType();
+    Type convViewElemType = getTypeConverter()->convertType(viewElemType);
+
+    // When the converted source operand element type is a struct, the original
+    // memref element type represents a SYCL type. Handle the non-SYCL case
+    // first.
+    if (!convSourceElemType.isa<LLVM::LLVMStructType>()) {
+      assert(
+          convViewElemType ==
+              prev.getType().cast<LLVM::LLVMPointerType>().getElementType() &&
+          "Expecting the element types to match");
+      auto memRefDesc = createMemRefDescriptor(
+          loc, viewMemRefType, targetMemRef.allocatedPtr(rewriter, loc),
+          rewriter.create<LLVM::GEPOp>(loc, prev.getType(), prev, idxs), sizes,
+          strides, rewriter);
+
+      rewriter.replaceOp(subViewOp, {memRefDesc});
       return success();
     }
 
-    assert(getTypeConverter()->convertType(viewMemRefType.getElementType()) ==
-               prev.getType().cast<LLVM::LLVMPointerType>().getElementType() &&
-           "Expecting the element types to match");
-    MemRefDescriptor nexRef = createMemRefDescriptor(
-        loc, viewMemRefType, targetMemRef.allocatedPtr(rewriter, loc),
-        rewriter.create<LLVM::GEPOp>(loc, prev.getType(), prev, idxs), sizes,
-        strides, rewriter);
-
-    rewriter.replaceOp(subViewOp, {nexRef});
+    // SYCL case
+    assert(sourceMemRefType.getRank() == viewMemRefType.getRank() &&
+           "Expecting the input and output MemRef ranks to be the same");
+    // Note:in MLIRASTConsumer::getMLIRType(), a memref of struct type is
+    // generated only for a struct that has at least one member of SYCL type,
+    // otherwise a llvm pointer type is generated instead of a memref).
+    assert((sycl::isSYCLType(sourceElemType) ||
+            sourceElemType.isa<LLVM::LLVMStructType>() &&
+                any_of(sourceElemType.cast<LLVM::LLVMStructType>().getBody(),
+                       [](const Type &memType) {
+                         return sycl::isSYCLType(memType);
+                       })) &&
+           "the source memref element type should be either a SYCL type, or "
+           "a struct containing at least a member of SYCL type");
+
+    SmallVector<Value, 4> indices;
+    computeIndices(convSourceElemType.cast<LLVM::LLVMStructType>(),
+                   convViewElemType, indices, subViewOp, transformed, rewriter);
+    assert(!indices.empty() && "Expecting a least one index");
+
+    // According to MLIRScanner::InitializeValueByInitListExpr() in
```

Comment:
Since you changed the other comment,
```suggestion
    // Note: MLIRScanner::InitializeValueByInitListExpr() in
```

Reply:
done